### ISSUE TITLE
Kernel dies sometimes when benchmark CSV.File() with 4 threads on Windows

### ISSUE THREAD

**Issue by strickek** [JuliaData/CSV.jl#984]
I'm not sure if the problem comes from `CSV` or `BenchmarkTools` or `Julia`. When I start Julia with 1 thread everything is ok. I discoverd this on IJulia but kann reproduce it in the REPL.

```
julia> versioninfo()
Julia Version 1.6.5
Commit 9058264a69 (2021-12-19 12:30 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: AMD A10-7800 Radeon R7, 12 Compute Cores 4C+8G
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, bdver3)
```

```
(@v1.6) pkg> st
      Status `C:\Users\Klaus\.julia\environments\v1.6\Project.toml`
  [6e4b80f9] BenchmarkTools v1.3.1
  [336ed68f] CSV v0.10.2
  [a93c6f00] DataFrames v1.3.2
  [7073ff75] IJulia v1.23.2
  [91a5bcdd] Plots v1.25.10
  [670cb119] ReadWriteDlm2 v0.8.2
  [9abbd945] Profile
  [10745b16] Statistics
```

`julia> Threads.nthreads()`
4

`julia> using DelimitedFiles, CSV, BenchmarkTools`

`julia> writedlm("test1.csv", rand(32161, 3))`

`julia> @btime CSV.File("test1.csv") # this line crashes in about 50%`

**Comment by quinnj** [comment]
I've dug into this for a while on a windows machine, and managed to reproduce at random a few times, but it doesn't seem to be related to any boundserror or other obvious problem with CSV.jl code specifically. I chatted with @keno about it and he suggested trying to run the code under wine (on linux?) in order to get an RR trace to debug; if anybody knows how to do that, it would be a great help and we can share with the main JuliaLang/julia repo and hopefully get some more internal eyes on the issue.

**Comment by nilshg** [comment]
This seems to be the issue I've been running into for some time in IJulia - although I'm surprised to hear it occurring in the REPL as well, I only ever saw it in IJulia. I would assume the relevance of `@btime` is just that it runs `CSV.File` multiple times, and that it crashes randomly so one could probably produce the same result by running `CSV.File` manually over and over again. 

I'd love to help with this and have both Windows and Linux machines available, but I confess I don't really understand how to implement Keno's suggestion - is the idea to download a Windows version of Julia and run this in wine on a Linux machine?

**Comment by davnn** [comment]
At least for me its very easy to reproduce on Windows with multithreading (unfortunately).

```julia
julia> versioninfo()
Julia Version 1.7.2
Commit bf53498635 (2022-02-06 15:21 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: AMD Ryzen 9 3900X 12-Core Processor

julia> Threads.nthreads()
24
```

The following loops crash at around 10 iterations.

```julia
using CSV, DataFrames

data = DataFrame(rand(10000, 100), :auto);
CSV.write("test.csv", data)

for i in 1:100
    @info "Opening CSV $i"
    CSV.File("test.csv")
end

for i in 1:100
    @info "Reading CSV $i"
    CSV.read("test.csv", DataFrame)
end
```

**Comment by nilshg** [comment]
Yes, getting CSV.read to crash isn't hard, my question was how to practically get an `rr` trace on Windows (as I've never used wine and wasn't sure what Keno suggested)

**Comment by davnn** [comment]
> Yes, getting CSV.read to crash isn't hard, my question was how to practically get an `rr` trace on Windows (as I've never used wine and wasn't sure what Keno suggested)

Unfortunately I cannot help in this regard. I tried some more debugging and it appears that the crash happens in
https://github.com/JuliaData/CSV.jl/blob/main/src/context.jl#L608.

**Comment by uwbanjoman** [comment]
```
julia> Threads.nthreads()
24

julia> versioninfo()
Julia Version 1.7.0
Commit 3bf9d17731 (2021-11-30 12:12 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: AMD Ryzen Threadripper 2920X 12-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, znver1)
```
I tried the above code several times on my machine and it never crashed.

**Comment by quinnj** [comment]
I'm going to close this as stale and hopefully fixed as of the recent 1.0 release. The internals were completely rewritten, so I'm hopeful random issues like this, along with improvements/fixes to Julia itself since 2022 have helped fix things. Comment/respond if you still see issues.
